Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift

```diff
@@ -0,0 +1,367 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+@_implementationOnly import OrderedCollections
+
+struct ChannelStateMachine<Element, Failure: Error>: Sendable {
+  private struct SuspendedProducer: Hashable {
+    let id: Int
+    let continuation: UnsafeContinuation<Void, Never>?
+    let element: Element?
+
+    func hash(into hasher: inout Hasher) {
+      hasher.combine(self.id)
+    }
+
+    static func == (_ lhs: SuspendedProducer, _ rhs: SuspendedProducer) -> Bool {
+      return lhs.id == rhs.id
+    }
+
+    static func placeHolder(id: Int) -> SuspendedProducer {
+      SuspendedProducer(id: id, continuation: nil, element: nil)
+    }
+  }
+
+  private struct SuspendedConsumer: Hashable {
+    let id: Int
+    let continuation: UnsafeContinuation<Element?, any Error>?
+
+    func hash(into hasher: inout Hasher) {
+      hasher.combine(self.id)
+    }
+
+    static func == (_ lhs: SuspendedConsumer, _ rhs: SuspendedConsumer) -> Bool {
+      return lhs.id == rhs.id
+    }
+
+    static func placeHolder(id: Int) -> SuspendedConsumer {
+      SuspendedConsumer(id: id, continuation: nil)
+    }
+  }
+
+  private enum Termination {
+    case finished
+    case failed(Error)
+  }
+
+  private enum State {
+    case channeling(
+      suspendedProducers: OrderedSet<SuspendedProducer>,
+      cancelledProducers: OrderedSet<SuspendedProducer>,
+      suspendedConsumers: OrderedSet<SuspendedConsumer>,
+      cancelledConsumers: OrderedSet<SuspendedConsumer>
+    )
+    case terminated(Termination)
+  }
+
+  private var state: State = .channeling(suspendedProducers: [], cancelledProducers: [], suspendedConsumers: [], cancelledConsumers: [])
+
+  enum SendAction {
+    case resumeConsumerAndExit(continuation: UnsafeContinuation<Element?, any Error>?)
+    case suspend
+  }
+
+  mutating func send() -> SendAction {
+    switch self.state {
+      case .channeling(_, _, let suspendedConsumers, _) where suspendedConsumers.isEmpty:
+        // we are idle or waiting for consumers, we have to suspend the producer
+        return .suspend
+
+      case .channeling(let suspendedProducers, let cancelledProducers, var suspendedConsumers, let cancelledConsumers):
+        // we are waiting for producers, we can resume the first available consumer
+        let suspendedConsumer = suspendedConsumers.removeFirst()
+        self.state = .channeling(
+          suspendedProducers: suspendedProducers,
+          cancelledProducers: cancelledProducers,
+          suspendedConsumers: suspendedConsumers,
+          cancelledConsumers: cancelledConsumers
+        )
+        return .resumeConsumerAndExit(continuation: suspendedConsumer.continuation)
+
+      case .terminated:
+        return .resumeConsumerAndExit(continuation: nil)
+    }
+  }
+
+  enum SendIsSuspendedAction {
+    case suspend
```

Comment:
hmm putting a case of `none` infers the type should really just not have that case and be optional.

to me `.suspended` seems fine.

Reply:
Yeah optional is fine as well. I really don't like suspended because we are already suspended here and nothing should really happen. We just need to wait for a consumer to show up.

Reply:
it is now an optional